### ISSUE TITLE
How to cache parts of an item?

### ISSUE THREAD

**Issue by bergus** [jaspervdj/hakyll#383]
Hi,
I've got some problems where Hakyll seems to recompile some pages even though their contents did not change. It is right that files they depend on did change, but not the parts (metadata) that are actually used.
Does Hakyll support this use case in its dependency tracking? Is there anything I can do to change my code (e.g. use `cached` somehow)?

Here's a very simplified example of how I create my menus:

``` hs
main = hakyll $ do
  match "pages/**.html" $ do
    route   idRoute
    compile $ do
      page <- getResourceBody
      saveSnapshot "raw" $ itemSetBody () page -- I thought deleting the body would help
      loadAndApplyTemplate "templates/default.html" pageContext page
  match "templates/*" $ compile templateCompiler

allPages :: Compiler [Item ()]
allPages = loadAllSnapshots ("pages/**.html") "raw"

pageContext = field "nav" makeMenu <> defaultContext

makeMenu :: Item String -> Compiler String
makeMenu curPage = do
  pages <- allPages
  menus <- mapM (loadAndApplyTemplate "templates/menuitem.html" menuItemContext) pages
  return $ intercalate "\n" menus
  where
    menuItemContext :: Context ()
    menuItemContext = boolField "active" isActive <> urlField "url" <> metadataField
    isActive menuItem = itemIdentifier curPage == itemIdentifier menuItem
```

My `default.html` template uses `$nav$` somewhere in a sidebar, and my `menuitem.html` template looks somewhat (without linebreaks, actually) like this:

``` html
$if(active)$
  <li class="active"><strong>$title$</strong></li>
$else$
  <li><a href="$url$">$title$</a></li>
$endif$
```

In reality, `pages` undergoes a lot more mangling, and the menu items are actually selected using `functionField`s in a template so that I can order them, and `allPages` does only load pages from a certain directory depending on which menu the page uses, but I think you get the gist. Disclaimer: Above code is untested.

Now when I change the `title` metadata of a page and call `build`, all pages are re-rendered and the menu is updated everywhere. However, if I change only the body of a page, it still **recompiles all pages not only the changed one**. I don't need to re-generate all menus. **How can I accomplish that?**

I guess `Hakyll.Web.Tags` has a similar problem, but I couldn't figure out whether or how it is solved there. Does collecting metadata in the `Rules` monad (and manually creating dependencies) make a difference here?

**Comment by jaspervdj** [comment]
The problem is `loadAllSnapshots`, which "requires" the body for every page. If you want to avoid this, try calling `getMatches` and then `getMetadata` for each page -- it might require some more manual plumbing though.

**Comment by bergus** [comment]
So if I call `getMetadata`, it'll only set up a dependency for the metadata part?
I'm just trying to understand the design (and my options within the current implementation) before I try to set up the manual plumbing.

**Comment by jaspervdj** [comment]
No, I think I was a bit unclear, sorry. There is no distinction between depending on just metadata or the whole thing -- although I would like to add it at some point. You need to do this directly in the `Rules` monad.
